### DISCUSSION TITLE
How do you handle errors that occur in middleware?

### DISCUSSION THREAD

**Discussion by ArtyomVancyan** [fastapi/fastapi#10404] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class OAuth2Middleware:
    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        try:
            await self.auth_middleware(scope, receive, send)
        except Exception as e:
            raise OAuth2Error(403, str(e))  # OAuth2Error inherits HTTPException
```

### Description

Well, I have a custom middleware where at the `self.auth_middleware(scope, receive, send)` step, an exception can occur. I want to catch and convert it to `HTTPException` so I can handle it in one module in the common way described in [this](https://fastapi.tiangolo.com/tutorial/handling-errors/) documentation. So, I have defined an error handler to catch the above exception:

```python
@app.exception_handler(OAuth2Error)
async def error_handler(request: Request, exc: OAuth2Error):
    return RedirectResponse(url="/", status_code=303)  # ignore the error and go to homepage
```

When the exception is raised from a router, this way of error handling works perfectly, but it does not work in my case when it is raised from a middleware. Should I do some hacks, such as changing the request context or the `scope` to achieve the expected behavior, or am I missing something?

### Operating System

Linux

### FastAPI Version

0.103.2

### Pydantic Version

2.4.2

### Python Version

3.10.7

**Comment by pcorvoh** [comment]
The processing of middleware in fastapi belongs to an onion model, so if you want to handle this error, you need to register a middleware in the outer layer to capture the error.
For example:
```python
class ExceptionMiddleware(BaseHTTPMiddleware):
     """Exception middleware"""

     async def dispatch(
             self,
             request: Request,
             call_next: RequestResponseEndpoint,
     ) -> Response:
         try:
             response: Response = await call_next(request)
         except (ValidationError, CancelledError, Exception,
                 OAuth2Middleware) as exc:
             if isinstance(exc, (SHTTP, FastapiHTTPException, OAuth2Middleware)):
                 err_msg: dict = exc.detail
                 content = {
                         "code": err_msg.get('code'),
                         "msg": err_msg.get('msg'),
                         "detail": {}
                 }
                 return JSONResponse(
                         status_code=exc.status_code, content=content
                 )
             else:
                 log.error(traceback.format_exc())
                 return JSONResponse(
                         content={
                                 "code": 500,
                                 "msg": "The programmer brother didn't get enough sleep and the system crashed!",
                                 "detail": {
                                         # "msg": "The server is abnormal, please contact the backend development!",
                                 }
                         },
         return response
```

**Reply by ArtyomVancyan** [reply]
Hi @pcorvoh, thanks for the answer, but I want to provide (rethrow/reraise) the exception above to handle it using the `@app.exception_handler(...)` decorator. [See](https://fastapi.tiangolo.com/tutorial/handling-errors/) how the exceptions are generally handled. I want it this way.

**Comment by thuyng-ing** [comment]
I have the same issue. Couldn't find the answer yet. Have you found the answer @ArtyomVancyan? Thanks!

**Reply by ArtyomVancyan** [reply]
No, did not.

**Reply by thuyng-ing** [reply]
I found someone has also exactly the same problem that we have on stackoverflow https://stackoverflow.com/questions/76092444/fastapi-exceptions-raised-in-middleware-are-not-handled-by-the-custom-exception

**Reply by thuyng-ing** [reply]
I have one workaround that could help you. Even though it doesn't go into my custom handler, it does go into a general handler 
` app.add_exception_handler(Exception, internal_error_exception_handler)`. In this general handler, you check for type and then execute handling based on the type, that should work. Hope this helps!

**Reply by thuyng-ing** [reply]
I also found out another way that you can create a new endpoint called exception, then you set ` request.scope['path'] = '/exception'` and set request.state with the info of the exception you need. Then in the newlywed created endpoint, you have a check and raise the corresponding exception. So, it will go to the correct exception handler. Those are the workaround ways I have tried. Hope this helps!
`

**Reply by thuyng-ing** [reply]
```
@app.middleware("http")
async def middleware_something(request: Request, call_next):
    try:
        a = something()
    except CustomException as e:
        request.scope['path'] = '/exception'
    request.state.exception = {"status_code": 401, "message": "some error"}
    response = await call_next(request)

@app.get("/exception")
async def exception_endpoint(request: Request):
    exception = request.state._state.get("exception")
    raise CustomException(status_code=exception.get("status_code"), message=exception.get("message"))
```
something like this. but this way it doesn't have the traces of the exception

**Comment by daliuchen** [comment]
FastAPI is implemented based on starlette. The processing of middleware belongs to an onion model.
The source code for building the middleware call stack: https://github.com/encode/starlette/blob/master/starlette/applications.py#L80.
The constructed call chain is:
> -> ServerErrorMiddleware -> user_middleware -> ExceptionMiddleware -> User Business Logic.

Moreover, in starlette, exceptions are divided into two types: Exception and Error.

   - Exception: represents known or expected exceptions, such as incorrect request parameters, non-existent users, etc. In my understanding, this represents business logic exceptions. In the constructed call chain, it is the layer closest to the user's business logic. The HTTPException provided by FastAPI is handled here. If HttpException is thrown in the user middleware, it will not take effect.
   - Error:  represents unknown exceptions, such as database connection failures, internal server errors, etc. Such errors will return a status code of 500.
   
   In the processing of starlette, Error has only one Handler (with an HTTP status code of 500 or when the handled error is an Exception), which is wrapped by ServerErrorMiddleware. While Exception has multiple Handlers and is wrapped by ExceptionMiddleware.
   
   My approach is to define a global Exception, have subclasses inherit from it, register a global ErrorHandler, throw an exception in the user middleware, and then handle the exception in the ErrorHandler and return it to the front end.
  
```python
# define base exception
class BaseException(Exception):
    def __init__(self, message: str, status_code: int):
        self.message = message
        self.status_code = status_code
        super().__init__(message)

    def content(self):
        return {"error_msg": self.message}

# define custom exception
class QuotaNotEnoughException(BaseException):
    def __init__(self):
        super().__init__("quota not enough", 403)


# register error handler
@fast_app.exception_handler(Exception)
async def unicorn_exception_handler(request: Request, exc: Exception):
    if isinstance(exc, BaseException):
        return JSONResponse(
            status_code=exc.status_code,
            content=exc.content(),
        )
    return JSONResponse(
        status_code=500,
        content={"error_msg": "Internal Server Error"},
    )

# user middleware
@fast_app.middleware("http")
async def auth_middleware(request: Request, call_next):
    if xxx:
        raise QuotaNotEnoughException()
    return await call_next(request)

```

**Comment by jovana** [comment]
I'm also struggling on the error handling raised from the middleware. 
I want a nice status code 403 and the details in the response. To simplify what I have done so fare examples below.

Normal using ```raise HTTPException(status_code=HTTP_403_FORBIDDEN, detail="ENDPOINT_ACCESS_DENIED")``` it always give the correct info to the client.

Raising above from the middleware ```add_middleware```, it always gives 500.

**Example of not working.**
RBAC middleware class:
```
class RBACMiddleware(BaseHTTPMiddleware):
    def __init__(self, app: FastAPI):
        super().__init__(app)

    async def dispatch(self, request: Request, call_next):
        request_method = str(request.method).upper()
        auth_header = request.headers.get("Authorization")
        raise HTTPException(status_code=HTTP_403_FORBIDDEN, detail="ENDPOINT_ACCESS_DENIED")
```

Add the RBACMiddleware to fastapi app:
```
app.add_middleware(RBACMiddleware)
```

In the above example it always raises the 500 internal server error:
```
Traceback (most recent call last):
  File "/Users/xxxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/uvicorn/protocols/http/h11_impl.py", line 406, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xxxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/fastapi/applications.py", line 1054, in __call__
    await super().__call__(scope, receive, send)
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/starlette/applications.py", line 113, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/starlette/middleware/errors.py", line 187, in __call__
    raise exc
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/starlette/middleware/errors.py", line 165, in __call__
    await self.app(scope, receive, _send)
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/starlette/middleware/base.py", line 185, in __call__
    with collapse_excgroups():
  File "/Users/xxx/.pyenv/versions/3.12.3/lib/python3.12/contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
    raise exc
  File "/Users/xxx/.pyenv/versions/3.12.3/envs/xxx-api/lib/python3.12/site-packages/starlette/middleware/base.py", line 187, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/app/application/auth/rbac.py", line 83, in dispatch
    raise HTTPException(status_code=HTTP_403_FORBIDDEN, detail="ENDPOINT_ACCESS_DENIED")
fastapi.exceptions.HTTPException: 403: ENDPOINT_ACCESS_DENIED
```

I have also tried to add a custom exception_handler:

```
@app.exception_handler(Exception)
async def exception_handler(request: Request, exc: Exception):
    print(f"type(exc): {type(exc)}")
    print(f"exc: {exc}")
    if isinstance(exc, HTTPException):
        return JSONResponse(
            status_code=status.HTTP_401_UNAUTHORIZED,
            content=jsonable_encoder({"detail": "ENDPOINT_ACCESS_DENIED", "body": exc.detail}),
        )
```

This still crashes the code but does not send the 500 response to the client, only there are no details/body set. Therefore it is hard to know what is going wrong on the client site. The response is not the same like using ```raise HTTPException(status_code=HTTP_403_FORBIDDEN, detail="ENDPOINT_ACCESS_DENIED")``` same where in the code.

**Reply by jovana** [reply]
I think I have a solution.
Also I have added a CORSMiddleware. It was inserted before. Now I have inserted it after (second one). This seems to solution.

```
app.add_middleware(
    CORSMiddleware,
    # allow_origins=origins,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)
```
Also changed the dispatch, and removed the raise statement.
Replaced like this

```
    async def dispatch(self, request: Request, call_next):
        request_method = str(request.method).upper()
        auth_header = request.headers.get("Authorization")
        return JSONResponse(
                  status_code=HTTP_403_FORBIDDEN,
                  content={"detail": "ENDPOINT_ACCESS_DENIED"}
                        )
```

**Comment by YuriiMotov** [comment, marked as answer]
From Starlette's docs: https://www.starlette.io/exceptions/#httpexception

> You should only raise HTTPException inside routing or endpoints. Middleware classes should instead just return appropriate responses directly.

Also: https://www.starlette.io/exceptions/#errors-and-handled-exceptions

> ... the middleware stack of a Starlette application is configured like this:
> * ServerErrorMiddleware - Returns 500 responses when server errors occur.
> * Installed middleware
> * ExceptionMiddleware - Deals with handled exceptions, and returns responses.
> * Router
> * Endpoints

So, exception handlers will not work for `HTTPExceptions` raised by middlewares, because they are raised outside of the `ExceptionMiddleware` which is in charge of this process.


As a solutions, you can:
* consider implementing your logic as a [custom router class](https://fastapi.tiangolo.com/how-to/custom-request-and-route/) instead of middleware
* implement the logic of exception handling using the specified exception handlers, similar to how it's done in `ExceptionMiddleware` (https://github.com/encode/starlette/blob/739321d719db5b6e3fff9ccca9a2ccc7c2e07f18/starlette/_exception_handler.py#L46-L53)
* hardcode the exception handlers in the middleware code
